### PR TITLE
Add ability to enable rmm pool on `cudf.pandas` import

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
This PR enables allocating of rmm memory pool on `cudf.pandas` import using the following environment variables:

```
export CUDF_PANDAS_RMM_MODE="pool"
```

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/pandas/__init__.py [hunk 1]
```diff
@@ -18,6 +18,25 @@ def install():
     loader = ModuleAccelerator.install("pandas", "cudf", "pandas")
     global LOADED
     LOADED = loader is not None
+    import os
+
+    cudf_pandas_mr = os.getenv("CUDF_PANDAS_MEMORY_RESOURCE", None)
+    if cudf_pandas_mr is not None:
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:24]:
```suggestion
    if cudf_pandas_mr := os.getenv("CUDF_PANDAS_MEMORY_RESOURCE", None) is not None:
```

File: python/cudf/cudf/pandas/__init__.py [hunk 2]
```diff
@@ -18,6 +18,25 @@ def install():
...
+        import rmm
+        import rmm.mr
+
+        cudf_pandas_mr = getattr(rmm.mr, cudf_pandas_mr, None)
+        if cudf_pandas_mr is not None:
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:29]: Use the walrus for brevity. Also, I think this should warn (or error?) if the specified attribute (the MR name) doesn't exist on `rmm.mr`, in an `else:` clause.
```suggestion
        if cudf_pandas_mr := getattr(rmm.mr, cudf_pandas_mr, None) is not None:
```

File: python/cudf/cudf/pandas/__init__.py [hunk 3]
```diff
@@ -18,6 +18,25 @@ def install():
     loader = ModuleAccelerator.install("pandas", "cudf", "pandas")
     global LOADED
     LOADED = loader is not None
+    import os
+
+    cudf_pandas_mr = os.getenv("CUDF_PANDAS_MEMORY_RESOURCE", None)
+    if cudf_pandas_mr is not None:
+        import rmm
+        import rmm.mr
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:26]: I think you only need the second one:
```suggestion
        import rmm.mr
```

**Review by bdice — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/pandas/__init__.py [hunk 4]
```diff
@@ -18,6 +18,25 @@ def install():
...
+
+        cudf_pandas_mr = getattr(rmm.mr, cudf_pandas_mr, None)
+        if cudf_pandas_mr is not None:
+            from rmm.mr import PoolMemoryResource
+
+            mr = PoolMemoryResource(
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:32]: Do we _always_ want the user's specified MR to be wrapped in a pool?

Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py:29]: I don't know, open to suggestions here. Or we can change this as we know which works the best from experimenting.

Comment (harrism) [reply python/cudf/cudf/pandas/__init__.py:29]: Certainly not. In the example in the PR description you specify CudeAsyncMemoryResource. That already has a pool, you probably wouldn't always want to wrap that in an RMM PoolMemoryResource.

I thought in the meeting we discussed just providing a limited set of choices here. More sophisticated users should be able to import RMM first and configure the current device resource to override if they have very specific MR needs.

**Review by harrism — CHANGES_REQUESTED** [review]
I don't think we should always wrap with a `PoolMemoryResource`

File: python/cudf/cudf/pandas/__init__.py (hunk 4 above)
Comment (bdice) [reply python/cudf/cudf/pandas/__init__.py:29]: We do a similar thing in libcudf NVBench benchmarks. Here are the options we expose there, for reference: https://github.com/rapidsai/cudf/blob/8cc4cc16fc6538f706b9f795d0879bdd0ba442a1/cpp/benchmarks/fixture/nvbench_fixture.hpp#L73-L84

Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py:29]: @harrism Do you feel comfortable having a similar option set for `cudf.pandas`?

Comment (harrism) [reply python/cudf/cudf/pandas/__init__.py:29]: Do you think we need arena?

Comment (madsbk) [reply python/cudf/cudf/pandas/__init__.py:29]: > Do you think we need arena?

Maybe, I haven't tried it, but its fragmentation avoidance might be useful. 
However, to limit the user complexity/documentation of cudf.pandas, I am also fine skipping `arena`. We can always add it later if it turns out to be useful.

Comment (harrism) [reply python/cudf/cudf/pandas/__init__.py:29]: I think you can also manually enable it by importing RMM and enabling it before importing cudf.pandas.

Comment (galipremsagar) [reply python/cudf/cudf/pandas/__init__.py]: I updated the PR to have simpler rmm mode selection and added pytests for the same.

@harrism @bdice This should be ready for review.

File: python/cudf/cudf/pandas/__init__.py [hunk 5]
```diff
@@ -18,6 +20,46 @@ def install():
     loader = ModuleAccelerator.install("pandas", "cudf", "pandas")
     global LOADED
     LOADED = loader is not None
+    import os
+
+    if (rmm_mode := os.getenv("CUDF_PANDAS_RMM_MODE", None)) is not None:
+        import rmm.mr
+        from rmm._lib.memory_resource import get_free_device_memory
```
Comment (galipremsagar) [inline python/cudf/cudf/pandas/__init__.py:27]: This method is being added in: https://github.com/rapidsai/rmm/pull/1567

**Review by harrism — APPROVED** [review]

File: python/cudf/cudf_pandas_tests/test_cudf_pandas.py [hunk 6]
```diff
@@ -1421,3 +1423,32 @@ def test_holidays_within_dates(holiday, start, expected):
...
+        capture_output=True,
+        text=True,
+        env=env,
+    )
+    if env_value in {"cuda", "pool", "async", "managed", "managed_pool"}:
+        assert sp_completed.returncode == 0
+    else:
+        assert sp_completed.returncode == 1
+
+    del env["CUDF_PANDAS_RMM_MODE"]
```
Comment (madsbk) [inline python/cudf/cudf_pandas_tests/test_cudf_pandas.py:1454]:
```suggestion
```

I don't you need this since `env` is a copy?

**Review by madsbk — APPROVED** [review]
Looks good

File: python/cudf/cudf/pandas/__init__.py [hunk 7]
```diff
@@ -19,6 +22,46 @@ def install():
     loader = ModuleAccelerator.install("pandas", "cudf", "pandas")
     global LOADED
     LOADED = loader is not None
+    import os
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:25]: Move this import to the top of the file.

File: python/cudf/cudf/pandas/__init__.py [hunk 8]
```diff
@@ -19,6 +22,46 @@ def install():
...
+
+    if (rmm_mode := os.getenv("CUDF_PANDAS_RMM_MODE", None)) is not None:
+        import rmm.mr
+        from rmm.mr import available_device_memory
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:29]: Should we import rmm at the top, too? I don’t think we have much benefit to lazy import here.

Also we should use the fully qualified name for rmm.mr.available_device_memory so we only have one import statement for rmm.mr.

File: python/cudf/cudf/pandas/__init__.py [hunk 9]
```diff
@@ -19,6 +22,46 @@ def install():
...
+            mr = rmm.mr.ManagedMemoryResource()
+            rmm.mr.set_current_device_resource(mr)
+        elif rmm_mode == "managed_pool":
+            rmm.reinitialize(
+                managed_memory=True,
+                pool_allocator=True,
+                initial_pool_size=free_memory,
+            )
+        else:
+            raise TypeError(f"Unsupported rmm mode: {rmm_mode}")
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:64]:
```suggestion
            raise ValueError(f"Unsupported rmm mode: {rmm_mode}")
```

File: python/cudf/cudf/pandas/__init__.py [hunk 10]
```diff
@@ -19,6 +22,46 @@ def install():
...
+                )
+            )
+        elif rmm_mode == "async":
+            mr = rmm.mr.CudaAsyncMemoryResource(initial_pool_size=free_memory)
+            rmm.mr.set_current_device_resource(mr)
+        elif rmm_mode == "managed":
+            mr = rmm.mr.ManagedMemoryResource()
+            rmm.mr.set_current_device_resource(mr)
+        elif rmm_mode == "managed_pool":
+            rmm.reinitialize(
```
Comment (bdice) [inline python/cudf/cudf/pandas/__init__.py:58]: Why does this call reinitialize while other modes create and set the current device resource?

**Review by bdice — APPROVED** [review]
A few suggestions/questions but none are blocking. Resolve as you see fit.

**Pull request merged**